tests/test_ssl_transport_close.py

```diff
@@ -0,0 +1,83 @@
+from aiohttp import web
+import aiohttp
+import asyncio
+import warnings
+
+
+async def handle_root(request):
+    """
+    respond after server was restarted, forcing connection_lost
+    """
+    cq = request.app['cq']
+    dq = request.app['cq']
+    i = await cq.get()
+    cq.task_done()
+    await dq.put(i)
+    return web.Response(text='')
+
+
+async def _prepare(aiohttp_server, ssl_ctx, client_ssl_ctx, cq, dq):
+    app = web.Application()
+    app['cq'] = cq
+    app['dq'] = dq
+    app.router.add_get('/', handle_root)
+    server = await aiohttp_server(app, port=12345, ssl=ssl_ctx)
+    asyncio.ensure_future(_restart(server.runner, ssl_ctx, cq))
+    session = aiohttp.ClientSession(connector=aiohttp.TCPConnector(ssl=client_ssl_ctx))
+    return server, session
+
+
+async def _close(server, session):
+    await server.close()
+    await session.close()
+
+async def _restart(runner, ssl_ctx, cq):
+    """
+    restart service to force connection_lost
+    """
+    while True:
+        for s in list(runner.sites):
+            await s.stop()
+            s = web.TCPSite(runner, '127.0.0.1', port=12345, ssl_context=ssl_ctx)
+            await s.start()
+            break
+        await cq.put(0)
+        await asyncio.sleep(0.5)
+
+def _ssl_resource_warnings(w):
+    unclosed_transport_msg = 'unclosed transport <asyncio.sslproto._SSLProtocolTransport object at'
+    return (
+        w.category == ResourceWarning
+        and w.filename.endswith('sslproto.py')
+        and w.message.args[0].startswith(unclosed_transport_msg)
+    )
+
+async def _run(loop, aiohttp_server, recwarn, ssl_ctx, client_ssl_ctx, cq, dq):
+    """
+    run for two processed client requests
+    """
+    server, session = await _prepare(aiohttp_server, ssl_ctx, client_ssl_ctx, cq, dq)
+
+    while dq.qsize() <= 2:
+        try:
+            await session.get('https://localhost:12345/')
+        except aiohttp.client.ClientConnectorError:
+            continue
+        finally:
+            await asyncio.sleep(0.1)
+
+        assert not len(list(filter(_ssl_resource_warnings, recwarn)))
+
+    await _close(server, session)
```

Comment:
Since this is not executed in a `finally:` block, it won't get executed in case of assertion errors potentially messing up external tests or anything that hardcodes its port to 12345.

Reply:
using port 0 now